core/drivers/crypto/se050/adaptors/utils/scp_config.c

```diff
@@ -402,34 +294,91 @@ static sss_status_t get_config_key(struct se050_scp_key *keys __maybe_unused)
 #endif
 }
 
-sss_status_t se050_scp03_get_keys(struct se050_scp_key *keys)
+sss_status_t se050_scp03_subkey_derive(struct se050_scp_key *keys)
 {
-	sss_status_t (*get_keys[])(struct se050_scp_key *) = {
-		&get_config_key, &get_db_key, &get_ofid_key,
-	};
-	size_t i = 0;
+	uint8_t msg[HUK_SUBKEY_MAX_LEN] = { 0 };
+	uint8_t key[HUK_SUBKEY_MAX_LEN] = { 0 };
+	uint8_t digest[sizeof(*keys)] = { 0 };
+	uint8_t *p = digest;
 
-	for (i = 0; i < ARRAY_SIZE(get_keys); i++)
-		if ((*get_keys[i])(keys) == kStatus_SSS_Success)
-			return kStatus_SSS_Success;
+	if (tee_otp_get_die_id(msg, HUK_SUBKEY_MAX_LEN))
+		return kStatus_SSS_Fail;
 
-	return kStatus_SSS_Fail;
+	if (huk_subkey_derive(HUK_SUBKEY_SE050, msg, HUK_SUBKEY_MAX_LEN,
+			      key, HUK_SUBKEY_MAX_LEN))
+		return kStatus_SSS_Fail;
+
+	if (tee_hash_createdigest(TEE_ALG_SHA384, key, HUK_SUBKEY_MAX_LEN,
```

Comment:
Here you're expanding 32 bytes to 48 bytes. Wouldn't it make sense to use `huk_subkey_derive()` three times instead, once for each element in `struct se050_scp_key` with a slightly altered `msg`?

Reply:
yes I thought of that but then I couldn't think of a benefit in terms of functionality/performance (am I missing something?); so I did what I think it is  a more efficient implementation: doesn't it provide the same level of security? or is it readability what you are concerned about?

Reply:
It is simplicity and that you add a dependency to another algorithm.

Reply:
@jenswi-linaro would something like this be acceptable?

trying to get agreement before pushing because every time I test I have to reprogram boards already using the previous keys which is a pain (there is no backwards compatibility obviously)...so I'd rather get it in the form you want before resubmitting the patch.

```
sss_status_t se050_scp03_subkey_derive(struct se050_scp_key *keys)
{
	uint8_t msg[SE050_SCP03_KEY_SZ] = { 0 };
	struct {
		const char *name;
		uint8_t *p;
	} key[3] = {
		[0] = { .name = "dek", .data = keys->dek },
		[1] = { .name = "mac", .data = keys->mac },
		[2] = { .name = "enc", .data = keys->enc },
	};
	size_t i = 0;

	if (tee_otp_get_die_id(msg, SE050_SCP03_KEY_SZ))
		return kStatus_SSS_Fail;

	for (i = 0; i < ARRAY_LEN(key); i++) {
		memcpy(msg, key[i].name, strlen(key[i].name));
		if (huk_subkey_derive(HUK_SUBKEY_SE050, msg, SE050_SCP03_KEY_SZ,
				      key[i].data, SE050_SCP03_KEY_SZ))
			return kStatus_SSS_Fail;
	}

	return kStatus_SSS_Success;
}
```

Reply:
It's fine to post untested patches in a situation like this. It's nice so say so then.
The code above likes nice and simple. You're overwriting the first part of the die id. Is that on purpose?

Reply:
ok. yes I am discarding the first part of the die_id just so all keys are different. but of course I am open to suggestions

Reply:
No big deal, I had expected concatenating the die id and the name of the key instead of discarding and replacing a part of the die id. It's bound to raise questions so please add short comment explaining why that choice was done.

Reply:
um, no I think you are right (better not to have two devices potentially using the same scp03 keys). ok will do that instead